src/models/StyledComponent.js

```diff
@@ -48,62 +51,111 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
       generatedClassName: '',
     }
 
+    constructor(props: Object, context: any) {
+      super(props, context)
+
+      let theme: any = null
+      let generatedClassName: string = ''
+
+      const styledContext = context[CHANNEL]
+      if (styledContext) {
+        const { getTheme } = styledContext
+        const contextTheme = getTheme()
+
+        // Props should take precedence over ThemeProvider, which should take precedence over
+        // defaultProps, but React automatically puts defaultProps on props.
+        const { defaultProps } = this.constructor
+        const isDefaultTheme = defaultProps && props.theme === defaultProps.theme
+        theme = props.theme && !isDefaultTheme ? props.theme : contextTheme
+        generatedClassName = this.generateAndInjectStyles(theme, props)
+      } else {
+        theme = props.theme || {}
+        generatedClassName = this.generateAndInjectStyles(
+          theme,
+          props,
+        )
+      }
+
+      this.state.theme = theme
+      this.state.generatedClassName = generatedClassName
+    }
+
     buildExecutionContext(theme: any, props: any) {
       const { attrs } = this.constructor
       const context = { ...props, theme }
       if (attrs === undefined) {
         return context
       }
 
-      this.attrs = Object.keys(attrs).reduce((acc, key) => {
+      const attrKeys = Object.keys(attrs)
+      for (let i = 0; i < attrKeys.length; i += 1) {
+        const key = attrKeys[i]
         const attr = attrs[key]
-        // eslint-disable-next-line no-param-reassign
-        acc[key] = typeof attr === 'function' ? attr(context) : attr
-        return acc
-      }, {})
+        if (typeof attr === 'function') {
+          attrs[key] = attr(context)
+        }
+      }
 
+      this.attrs = attrs
       return { ...context, ...this.attrs }
     }
 
     generateAndInjectStyles(theme: any, props: any) {
-      const { componentStyle, warnTooManyClasses } = this.constructor
+      const { componentStyle } = this.constructor
+      if (componentStyle.isStaticRules) {
+        // Do fancy static checks to avoid class name generation
+        if (componentStyle.staticGeneratedClassName) {
```

Comment:
Since you are already branching inside `ComponentStyle` this is both duplicated logic, and creates a dead branch inside the `ComponentStyle` instance.

Unintuitively, not introducing the branch here should be faster since you'd then always call the same method, regardless of the input, which should keep v8's optimiser happy.

Revert is preferable :)